internal/sfdisk/sfdisk.go

```diff
@@ -0,0 +1,276 @@
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os/exec"
+
+	"github.com/coreos/ignition/v2/config/util"
+	"github.com/coreos/ignition/v2/config/v3_5_experimental/types"
+	"github.com/coreos/ignition/v2/internal/distro"
+	"github.com/coreos/ignition/v2/internal/log"
+)
+
+type Operation struct {
+	logger    *log.Logger
+	dev       string
+	wipe      bool
+	parts     []Partition
+	deletions []int
+	infos     []int
+}
+
+// We ignore types.Partition.StartMiB/SizeMiB in favor of
+// StartSector/SizeInSectors.  The caller is expected to do the conversion.
+type Partition struct {
+	types.Partition
+	StartSector   *int64
+	SizeInSectors *int64
+
+	// shadow StartMiB/SizeMiB so they're not accidentally used
+	StartMiB string
+	SizeMiB  string
+}
+
+// Begin begins an sfdisk operation
+func Begin(logger *log.Logger, dev string) *Operation {
+	return &Operation{logger: logger, dev: dev}
+}
+
+// CreatePartition adds the supplied partition to the list of partitions to be created as part of an operation.
+func (op *Operation) CreatePartition(p Partition) {
+	op.parts = append(op.parts, p)
+}
+
+func (op *Operation) DeletePartition(num int) {
+	op.deletions = append(op.deletions, num)
+}
+
+func (op *Operation) Info(num int) {
+	op.infos = append(op.infos, num)
+}
+
+// WipeTable toggles if the table is to be wiped first when committing this operation.
+func (op *Operation) WipeTable(wipe bool) {
+	op.wipe = wipe
+}
+
+// Pretend is like Commit() but uses the --no-act flag and returns the output
+func (op *Operation) Pretend() (string, error) {
+	// Handle deletions first
+	if err := op.handleDeletions(); err != nil {
+		return "", err
+	}
+
+	if err := op.handleInfo(); err != nil {
+		return "", err
+	}
+
+	script := op.sfdiskBuildOptions()
+	cmd := exec.Command("sh", "-c", fmt.Sprintf("echo -e \"%s\" | sudo %s --no-act %s", script, distro.SfdiskCmd(), op.dev))
+	stdout, err := cmd.StdoutPipe()
+
+	if err != nil {
+		return "", err
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return "", err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return "", err
+	}
+	output, err := io.ReadAll(stdout)
+	if err != nil {
+		return "", err
+	}
+
+	errors, err := io.ReadAll(stderr)
+	if err != nil {
+		return "", err
+	}
+
+	if err := cmd.Wait(); err != nil {
+		return "", fmt.Errorf("failed to pretend to create partitions. Err: %v. Stderr: %v", err, string(errors))
+	}
+
+	return string(output), nil
+}
+
+// Commit commits an partitioning operation.
+func (op *Operation) SfdiskCommit() error {
+	script := op.sfdiskBuildOptions()
+	if len(script) == 0 {
+		return nil
+	}
+
+	// If wipe we need to reset the partition table
+	if op.wipe {
+		// sgdisk --zap-all /dev/sda destroys the partition table, let's do the same with wipefs
+		// Found out that wipefs is not sufficient, as it results in a corrupt partition table, look into sgdisk's way of doing --zap-all
+		cmd := exec.Command("sudo", distro.WipefsCmd(), "-t pmbr", "-t mbr", "-t gpt", "-a", op.dev)
```

Comment:
```suggestion
		cmd := exec.Command("sudo", distro.WipefsCmd(), "-t", "pmbr,mbr,gpt", "-a", op.dev)
```

Though as per https://github.com/coreos/ignition/pull/1929/files#r1757502849 I'm starting to think we don't actually need `-t` and it'll do the right thing for us. By comparison, if I have an XFS filesystem directly on `/dev/vda` (not a partition) and I do `sgdisk --zap-all /dev/vda`, it'll happily nuke that data.

So then this would just be

```suggestion
		cmd := exec.Command("sudo", distro.WipefsCmd(), "-a", op.dev)
```

Reply:
100%, makes sense. Thank you for looking into that.